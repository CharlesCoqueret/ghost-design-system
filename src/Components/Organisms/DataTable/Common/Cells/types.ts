import {
  IColumnType,
  IExtraEditableDataTableProps,
  IExtraLineEditableInPlaceDataTableProps,
  IExtraStaticDataTableProps,
} from '../types';

export interface ICellProps<T, U extends IColumnType<T>> {
  column: U;
  dataTestId?: string;
  editing?: boolean;
  extra?: IExtraStaticDataTableProps<T> | IExtraLineEditableInPlaceDataTableProps<T> | IExtraEditableDataTableProps<T>;
  forcedValue?: T[keyof T] | number | string | Date | null;
  onChange?: (newValue: T[keyof T]) => void;
  row?: T;
  rowIndex: number;
}
