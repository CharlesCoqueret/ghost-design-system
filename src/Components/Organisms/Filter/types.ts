import { ThousandsGroupStyle } from '../../Atoms/AmountInput';
import { DateFormat, WeekDayEnum } from '../../Atoms/DatePickerInput';
import { IOption } from '../../Atoms/SelectInput';

export enum FilterTypeEnum {
  CHECKBOX = 'checkbox',
  DATE = 'date',
  DYNAMICSEARCH = 'dynamicsearch',
  MULTISELECT = 'multiselect',
  NUMBER = 'number',
  COLUMN = 'column',
  SELECT = 'select',
  TEXT = 'text',
  TITLE = 'title',
}

export type IFilterLayoutAndFieldsProps<T> = IFilerColProps<T> | IFilerTitleProps | IFilterFieldsProps<T>;

export type IFilterFieldsProps<T> =
  | IFilterCheckboxProps<T>
  | IFilterDateProps<T>
  | IFilterDynamicSearchProps<T>
  | IFilterMultiSelectProps<T>
  | IFilterNumberProps<T>
  | IFilterSelectProps<T>
  | IFilterTextProps<T>;

export interface IFilterBaseProps<T> {
  dataIndex: keyof T;
  dataTestId?: string;
  fieldClassName?: string;
  fieldSize?: number;
  helperText?: string;
  inputClassName?: string;
  label?: string;
}

export interface IFilterCheckboxProps<T> extends IFilterBaseProps<T> {
  filterType: FilterTypeEnum.CHECKBOX;
}

export interface IFilterDateProps<T> extends IFilterBaseProps<T> {
  calendarStartDay?: WeekDayEnum;
  dateFormat?: DateFormat;
  filterType: FilterTypeEnum.DATE;
  isClearable?: boolean;
  locale?: string;
  name?: string;
  placeholder?: string;
}

export interface IFilterDynamicSearchProps<T> extends IFilterBaseProps<T> {
  colors?: {
    controlErrorColor: string; // colors.error,
    controlFocusColor: string; // colors.primary,
    fontColor: string; // 'rgb(0, 0, 0)',
    optionFocusColor: string; // colors.chalk,
    optionSelectedColor: string; // colors.primary,
  };
  filterType: FilterTypeEnum.DYNAMICSEARCH;
  isClearable?: boolean;
  name?: string;
  noOptionsMessage: (obj: { inputValue: string }) => string;
  placeholder?: string;
  resolveValue: (value: string | number) => Promise<IOption | undefined>;
  searchOptions: (searchTerm: string) => Promise<Array<IOption> | undefined>;
}

export interface IFilterMultiSelectProps<T> extends IFilterBaseProps<T> {
  colors?: {
    controlErrorColor: string; // colors.error,
    controlFocusColor: string; // colors.primary,
    fontColor: string; // 'rgb(0, 0, 0)',
    optionFocusColor: string; // colors.chalk,
    optionSelectedColor: string; // colors.primary,
  };
  filterType: FilterTypeEnum.MULTISELECT;
  isClearable?: boolean;
  name?: string;
  numberOfItemLabel: string;
  numberOfItemsLabel: string;
  options: Array<IOption>;
  placeholder?: string;
}

export interface IFilterNumberProps<T> extends IFilterBaseProps<T> {
  allowNegative?: boolean;
  decimalScale?: number;
  decimalSeparator?: string;
  filterType: FilterTypeEnum.NUMBER;
  maxValue?: number;
  minValue?: number;
  name?: string;
  placeholder?: string;
  prefix?: string | undefined;
  suffix?: string | undefined;
  thousandSeparator?: string;
  thousandsGroupStyle?: ThousandsGroupStyle;
}

export interface IFilerColProps<T> {
  filterType: FilterTypeEnum.COLUMN;
  fields: Array<IFilterLayoutAndFieldsProps<T>>;
}

export interface IFilterSelectProps<T> extends IFilterBaseProps<T> {
  colors?: {
    controlErrorColor: string; // colors.error,
    controlFocusColor: string; // colors.primary,
    fontColor: string; // 'rgb(0, 0, 0)',
    optionFocusColor: string; // colors.chalk,
    optionSelectedColor: string; // colors.primary,
  };
  filterType: FilterTypeEnum.SELECT;
  isClearable?: boolean;
  name?: string;
  options: Array<IOption>;
  placeholder?: string;
}

export interface IFilterTextProps<T> extends IFilterBaseProps<T> {
  filterType: FilterTypeEnum.TEXT;
  maxLength?: number;
  minLength?: number;
  name?: string;
  placeholder?: string;
}

export interface IFilerTitleProps {
  filterType: FilterTypeEnum.TITLE;
  label: string;
}
