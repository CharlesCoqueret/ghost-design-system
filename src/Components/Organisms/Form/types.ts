import { ReactElement } from 'react';

import { IOption } from '../../Atoms/SelectInput';
import { IAmountFieldProps } from '../../Molecules/AmountField/AmountField';
import { IDatePickerFieldProps } from '../../Molecules/DatePickerField/DatePickerField';
import { IFileFieldProps } from '../../Molecules/FileField/FileField';
import { IDynamicSearchFieldProps } from '../../Molecules/DynamicSearchField/DynamicSearchField';
import { IMultiSelectFieldProps } from '../../Molecules/MultiSelectField/MultiSelectField';
import { IPercentageFieldProps } from '../../Molecules/PercentageField/PercentageField';
import { IRichTextFieldProps } from '../../Molecules/RichTextField/RichTextField';
import { ISelectFieldProps } from '../../Molecules/SelectField/SelectField';
import { ILineEditableDataTableProps } from '../DataTable/LineEditableDataTable/LineEditableDataTable';
import { ITextFieldProps } from '../../Molecules/TextField/TextField';
import { ITextAreaFieldProps } from '../../Molecules/TextAreaField/TextAreaField';
import { IYearPickerFieldProps } from '../../Molecules/YearPickerField/YearPickerField';

export interface IFormSubmitReturnedType<T> {
  data: T;
  valid: boolean;
}

export interface IUseFormReturnedType<T> {
  formElement: ReactElement;
  getData: () => T;
  isModified: () => boolean;
  reset: () => void;
  submit: () => IFormSubmitReturnedType<T>;
}

export enum FieldTypeEnum {
  AMOUNT = 'amount',
  CHECKBOX = 'checkbox',
  CUSTOM = 'custom',
  DATE = 'date',
  DESCRIPTION = 'description',
  DYNAMICSEARCH = 'dynamicsearch',
  FILE = 'file',
  MULTISELECT = 'multiselect',
  NUMBER = 'number',
  PERCENTAGE = 'percentage',
  RICHTEXT = 'richtext',
  SECTION = 'section',
  SELECT = 'select',
  SWITCH = 'switch',
  TABLE = 'table',
  TEXT = 'text',
  TEXTAREA = 'textarea',
  YEAR = 'year',
}

export type IFieldAndLayoutProps<T> = IFieldProps<T> | ILayoutProps<T>;

export type ILayoutProps<T> = IFieldSectionProps<T> | IFieldDescriptionProps<T>;

export type IFieldProps<T> =
  | IFieldAmountProps<T>
  | IFieldCheckboxProps<T>
  | IFieldCustomProps<T>
  | IFieldDateProps<T>
  | IFieldDynamicSearchProps<T>
  | IFieldFileProps<T>
  | IFieldMultiSelectProps<T>
  | IFieldNumberProps<T>
  | IFieldPercentageProps<T>
  | IFieldRichtextProps<T>
  | IFieldSelectProps<T>
  | IFieldSwitchProps<T>
  // TODO investigate type resolution
  // Using any to avoid circular type definition for now, until there is a way to get the type of an item of T[keyof T]
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  | IFieldTableProps<T, any>
  | IFieldTextAreaProps<T>
  | IFieldTextProps<T>
  | IFieldYearProps<T>;

export interface IVisibilityProps<T> {
  /** When set to true or when called, returning true, it hides the field or description or section (and children fields) (optional, default: false) */
  hidden?: boolean | ((data: T) => boolean);
}

export interface IFieldBaseProps<T> extends Partial<IVisibilityProps<T>> {
  dataIndex: keyof T;
  disabled?: boolean;
  errorMessage?: string;
  fieldClassName?: string;
  helperText?: string;
  highlighted?: boolean;
  inputClassName?: string;
  label: string;
  mandatory?: boolean;
  name?: string;
  readOnly?: boolean;
}

export interface IFieldAmountProps<T>
  extends IFieldBaseProps<T>,
    Pick<
      IAmountFieldProps,
      | 'allowNegative'
      | 'decimalScale'
      | 'decimalSeparator'
      | 'maxValue'
      | 'minValue'
      | 'placeholder'
      | 'prefix'
      | 'suffix'
      | 'thousandSeparator'
      | 'thousandsGroupStyle'
    > {
  fieldType: FieldTypeEnum.AMOUNT;
}

export interface IFieldCheckboxProps<T> extends IFieldBaseProps<T> {
  fieldType: FieldTypeEnum.CHECKBOX;
}

export interface IFieldCustomProps<T, U = unknown> extends IFieldBaseProps<T> {
  fieldType: FieldTypeEnum.CUSTOM;
  isEqual?: (previousValue: T[keyof T], currentValue: T[keyof T]) => boolean;
  data?: U;
  customField: <
    U extends {
      highlighted?: boolean;
      onChange?: (value: T[keyof T]) => void;
      readOnly?: boolean;
      inputValue?: T[keyof T];
    },
  >(
    props: U,
  ) => ReactElement;
}

export interface IFieldDateProps<T>
  extends IFieldBaseProps<T>,
    Pick<IDatePickerFieldProps, 'calendarStartDay' | 'dateFormat' | 'isClearable' | 'locale' | 'placeholder'> {
  fieldType: FieldTypeEnum.DATE;
}

export interface IFieldDescriptionProps<T> extends Partial<IVisibilityProps<T>> {
  dataIndex?: keyof T;
  description: ReactElement | (<U extends { value: T[keyof T] }>(props: U) => ReactElement);
  fieldType: FieldTypeEnum.DESCRIPTION;
}

export interface IFieldFileProps<T>
  extends IFieldBaseProps<T>,
    Pick<
      IFileFieldProps,
      | 'acceptTypes'
      | 'additionalInfo'
      | 'maxFiles'
      | 'maxFileSize'
      | 'maxFolderDepth'
      | 'onDelete'
      | 'onDownload'
      | 'requestHeaders'
      | 'requestMethod'
      | 'requestUrl'
      | 'requestWithCredentials'
      | 'showFileSize'
      | 'showProgressBar'
      | 'uploadMessage'
      | 'localization'
    > {
  fieldType: FieldTypeEnum.FILE;
}

export interface IFieldDynamicSearchProps<T>
  extends IFieldBaseProps<T>,
    Pick<
      IDynamicSearchFieldProps,
      'colors' | 'isClearable' | 'noOptionsMessage' | 'placeholder' | 'resolveValue' | 'searchOptions'
    > {
  fieldType: FieldTypeEnum.DYNAMICSEARCH;
}

export interface IFieldMultiSelectProps<T>
  extends IFieldBaseProps<T>,
    Pick<
      IMultiSelectFieldProps,
      'colors' | 'isClearable' | 'numberOfItemLabel' | 'numberOfItemsLabel' | 'placeholder'
    > {
  // When the value is not present in the options, should the value be erased (optional, default: false)
  eraseValueWhenNotInOptions?: boolean;
  fieldType: FieldTypeEnum.MULTISELECT;
  options: Array<IOption> | ((data: T) => Array<IOption>);
}

export interface IFieldNumberProps<T>
  extends IFieldBaseProps<T>,
    Pick<
      IAmountFieldProps,
      | 'allowNegative'
      | 'decimalScale'
      | 'decimalSeparator'
      | 'maxValue'
      | 'minValue'
      | 'placeholder'
      | 'prefix'
      | 'suffix'
      | 'thousandSeparator'
      | 'thousandsGroupStyle'
    > {
  fieldType: FieldTypeEnum.NUMBER;
}

export interface IFieldPercentageProps<T>
  extends IFieldBaseProps<T>,
    Pick<
      IPercentageFieldProps,
      | 'allowNegative'
      | 'decimalScale'
      | 'decimalSeparator'
      | 'maxValue'
      | 'minValue'
      | 'placeholder'
      | 'thousandSeparator'
      | 'thousandsGroupStyle'
    > {
  fieldType: FieldTypeEnum.PERCENTAGE;
}

export interface IFieldRichtextProps<T>
  extends IFieldBaseProps<T>,
    Pick<IRichTextFieldProps, 'convertImagesToBase64' | 'enableImage' | 'enableLink' | 'locale' | 'maxLength'> {
  fieldType: FieldTypeEnum.RICHTEXT;
}

export interface IFieldSectionProps<T> extends Partial<IVisibilityProps<T>> {
  collapsable?: boolean;
  fieldType: FieldTypeEnum.SECTION;
  openInitially?: boolean;
  fields: Array<IFieldProps<T>>;
  label: string;
}

export interface IFieldSelectProps<T>
  extends IFieldBaseProps<T>,
    Pick<ISelectFieldProps, 'colors' | 'isClearable' | 'placeholder'> {
  // When the value is not present in the options, should the value be erased (optional, default: false)
  eraseValueWhenNotInOptions?: boolean;
  fieldType: FieldTypeEnum.SELECT;
  options: Array<IOption> | ((data: T) => Array<IOption>);
}

export interface IFieldSwitchProps<T> extends IFieldBaseProps<T> {
  fieldType: FieldTypeEnum.SWITCH;
}

export interface IFieldTableProps<T, U>
  extends IFieldBaseProps<T>,
    Pick<ILineEditableDataTableProps<U>, 'columns' | 'extra' | 'loading' | 'onSortChange'> {
  fieldType: FieldTypeEnum.TABLE;
}

export interface IFieldTextProps<T>
  extends IFieldBaseProps<T>,
    Pick<ITextFieldProps, 'maxLength' | 'minLength' | 'placeholder'> {
  fieldType: FieldTypeEnum.TEXT;
}

export interface IFieldTextAreaProps<T>
  extends IFieldBaseProps<T>,
    Pick<ITextAreaFieldProps, 'maxLength' | 'minLength' | 'placeholder'> {
  fieldType: FieldTypeEnum.TEXTAREA;
}

export interface IFieldYearProps<T> extends IFieldBaseProps<T>, Pick<IYearPickerFieldProps, 'placeholder'> {
  fieldType: FieldTypeEnum.YEAR;
}
