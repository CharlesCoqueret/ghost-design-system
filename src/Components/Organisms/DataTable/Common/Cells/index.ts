export { default as AmountCell } from './AmountCell';
export { default as BadgeCell } from './BadgeCell';
export { default as ButtonCell } from './ButtonCell';
export { default as CheckboxCell } from './CheckboxCell';
export { default as CodeCell } from './CodeCell';
export { default as CustomCell } from './CustomCell';
export { default as DateCell } from './DateCell';
export { default as FileCell } from './FileCell';
export { default as DescriptionCell } from './DescriptionCell';
export { default as DynamicSearchCell } from './DynamicSearchCell';
export { default as MultiSelectCell } from './MultiSelectCell';
export { default as NumberCell } from './NumberCell';
export { default as PercentageCell } from './PercentageCell';
export { default as RichTextCell } from './RichTextCell';
export { default as SectionCell } from './SectionCell';
export { default as SwitchCell } from './SwitchCell';
export { default as TableCell } from './TableCell';
export { default as TextCell } from './TextCell';
export { default as TextAreaCell } from './TextAreaCell';
export { default as YearCell } from './YearCell';
export type { ICellProps } from './types';
