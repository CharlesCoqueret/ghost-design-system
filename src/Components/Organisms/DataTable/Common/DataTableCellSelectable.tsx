import React, { KeyboardEvent, MouseEvent } from 'react';
import classnames from 'classnames';

import { Icon } from '../../../Atoms/Icon';

interface IDataTableCellSelectableProps {
  handleSelectClick?: (event: MouseEvent<HTMLElement> | KeyboardEvent<HTMLElement>, selected: boolean) => void;
  selected?: boolean;
  selectable: boolean;
  dataTestId?: string;
}

interface IDataTableCellSelectableState {
  selected: boolean;
}

/**
 * DataTableCellSelectable creates a table cell containing a checkbox.
 * Note: It has been build as a class component to be able to bind the onClick action to this component only, avoiding other component to catch this click.
 */
class DataTableCellSelectable extends React.Component<IDataTableCellSelectableProps, IDataTableCellSelectableState> {
  constructor(props: IDataTableCellSelectableProps) {
    super(props);
    this.state = { selected: this.props.selected || false };

    this.onSelect = this.onSelect.bind(this);
  }

  onSelect(event: MouseEvent<HTMLElement> | KeyboardEvent<HTMLElement>): void {
    if (event.type === 'keyup' && (event as KeyboardEvent).key !== 'Enter') {
      return;
    }

    event.stopPropagation();

    const selected = !this.props.selected;
    this.setState({ ...this.state, selected: selected });
    if (this.props.handleSelectClick) {
      this.props.handleSelectClick(event, selected);
    } else {
      console.error('Missing handleSelectClick');
    }
  }

  render() {
    return (
      <td
        key='cell-selectable'
        className={this.props.selectable ? 'table--value--selectable' : 'table--value--selectable-disabled'}
        // eslint-disable-next-line @typescript-eslint/unbound-method
        onClick={this.props.selectable ? this.onSelect : undefined}
        // eslint-disable-next-line @typescript-eslint/unbound-method
        onKeyUp={this.props.selectable ? this.onSelect : undefined}
        tabIndex={this.props.selectable ? 0 : -1}>
        <div className={classnames('checkbox-marker', { selected: this.props.selected })}>
          <Icon
            icon={[
              this.props.selected || !this.props.selectable ? 'fas' : 'fal',
              this.props.selected ? 'square-check' : 'square',
            ]}
            size='lg'
            data-testid={this.props.dataTestId}
          />
        </div>
      </td>
    );
  }
}

export default DataTableCellSelectable;
