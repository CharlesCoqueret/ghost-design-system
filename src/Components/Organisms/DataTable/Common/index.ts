export { ColumnType, SortDirectionEnum } from './types';
export type {
  IButtonCellProps,
  IColumnType,
  IExtraEditableDataTableProps,
  IExtraLineEditableInPlaceDataTableProps,
  IExtraStaticDataTableProps,
  TableType,
} from './types';
