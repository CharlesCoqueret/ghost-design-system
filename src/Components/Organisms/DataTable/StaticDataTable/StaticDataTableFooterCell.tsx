import React, { ReactElement } from 'react';

import { ColumnType, IColumnType, IExtraStaticDataTableProps } from '../Common/types';
import { AmountCell, NumberCell, PercentageCell } from '../Common/Cells';

export interface IStaticDataTableFooterCellProps<T> {
  column: IColumnType<T>;
  data: Array<T>;
  extra?: IExtraStaticDataTableProps<T>;
  rowIndex: number;
}

const StaticDataTableFooterCell = <T,>(props: IStaticDataTableFooterCellProps<T>): ReactElement => {
  const { column, data, extra, rowIndex } = props;

  switch (column.type) {
    case ColumnType.AMOUNT: {
      return (
        <AmountCell<T>
          column={column}
          forcedValue={extra?.computeTotal && extra.computeTotal(data, column.dataIndex)}
          extra={extra}
          rowIndex={rowIndex}
        />
      );
    }
    case ColumnType.NUMBER: {
      return (
        <NumberCell<T>
          column={column}
          forcedValue={extra?.computeTotal && extra.computeTotal(data, column.dataIndex)}
          extra={extra}
          rowIndex={rowIndex}
        />
      );
    }
    case ColumnType.PERCENTAGE: {
      return (
        <PercentageCell<T>
          column={column}
          forcedValue={extra?.computeTotal && extra.computeTotal(data, column.dataIndex)}
          extra={extra}
          rowIndex={rowIndex}
        />
      );
    }
    case ColumnType.BADGE:
    case ColumnType.BUTTON:
    case ColumnType.CODE:
    case ColumnType.CUSTOM:
    case ColumnType.DATE:
    case ColumnType.TEXT:
      return <td className='table--footer-value center'>-</td>;

    default: {
      throw new Error('Missing ColumnType');
    }
  }

  throw new Error('Should have returned by then');
};

export default StaticDataTableFooterCell;
