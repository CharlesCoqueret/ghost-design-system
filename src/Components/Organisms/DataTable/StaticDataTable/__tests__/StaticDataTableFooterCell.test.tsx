import React from 'react';
import { render } from '@testing-library/react';
import { ColumnType } from '../../Common/types';

import StaticDataTableFooterCell from '../StaticDataTableFooterCell';

describe('StaticDataTableFooterCell component', () => {
  it('StaticDataTableFooterCell renders with amount', () => {
    const computeTotalMock = jest.fn();

    const { container, rerender } = render(
      <table>
        <tbody>
          <tr>
            <StaticDataTableFooterCell<{ amount: number }>
              column={{
                dataIndex: 'amount',
                title: 'Amount',
                type: ColumnType.AMOUNT,
              }}
              data={[{ amount: 1 }, { amount: 2 }]}
              extra={{ computeTotal: computeTotalMock }}
              rowIndex={3}
            />
          </tr>
        </tbody>
      </table>,
    );

    expect(container).toMatchSnapshot();
    expect(computeTotalMock).toBeCalledTimes(1);
    expect(computeTotalMock).toBeCalledWith([{ amount: 1 }, { amount: 2 }], 'amount');

    rerender(
      <table>
        <tbody>
          <tr>
            <StaticDataTableFooterCell<{ amount: number }>
              column={{
                dataIndex: 'amount',
                title: 'Amount',
                type: ColumnType.AMOUNT,
              }}
              data={[{ amount: 1 }, { amount: 2 }]}
              rowIndex={3}
            />
          </tr>
        </tbody>
      </table>,
    );
    expect(container).toMatchSnapshot();
  });

  it('StaticDataTableFooterCell renders with number', () => {
    const computeTotalMock = jest.fn();

    const { container, rerender } = render(
      <table>
        <tbody>
          <tr>
            <StaticDataTableFooterCell<{ number: number }>
              column={{
                dataIndex: 'number',
                title: 'Number',
                type: ColumnType.NUMBER,
              }}
              data={[{ number: 1 }, { number: 2 }]}
              extra={{ computeTotal: computeTotalMock }}
              rowIndex={3}
            />
          </tr>
        </tbody>
      </table>,
    );

    expect(container).toMatchSnapshot();
    expect(computeTotalMock).toBeCalledTimes(1);
    expect(computeTotalMock).toBeCalledWith([{ number: 1 }, { number: 2 }], 'number');

    rerender(
      <table>
        <tbody>
          <tr>
            <StaticDataTableFooterCell<{ number: number }>
              column={{
                dataIndex: 'number',
                title: 'Number',
                type: ColumnType.NUMBER,
              }}
              data={[{ number: 1 }, { number: 2 }]}
              rowIndex={3}
            />
          </tr>
        </tbody>
      </table>,
    );

    expect(container).toMatchSnapshot();
  });

  it('StaticDataTableFooterCell renders with percentage', () => {
    const computeTotalMock = jest.fn();

    const { container, rerender } = render(
      <table>
        <tbody>
          <tr>
            <StaticDataTableFooterCell<{ percentage: number }>
              column={{
                dataIndex: 'percentage',
                title: 'Percentage',
                type: ColumnType.PERCENTAGE,
              }}
              data={[{ percentage: 1 }, { percentage: 2 }]}
              extra={{ computeTotal: computeTotalMock }}
              rowIndex={3}
            />
          </tr>
        </tbody>
      </table>,
    );

    expect(container).toMatchSnapshot();
    expect(computeTotalMock).toBeCalledTimes(1);
    expect(computeTotalMock).toBeCalledWith([{ percentage: 1 }, { percentage: 2 }], 'percentage');

    rerender(
      <table>
        <tbody>
          <tr>
            <StaticDataTableFooterCell<{ percentage: number }>
              column={{
                dataIndex: 'percentage',
                title: 'Percentage',
                type: ColumnType.PERCENTAGE,
              }}
              data={[{ percentage: 1 }, { percentage: 2 }]}
              rowIndex={3}
            />
          </tr>
        </tbody>
      </table>,
    );

    expect(container).toMatchSnapshot();
  });

  it('StaticDataTableFooterCell renders with badge, button, checkbox, code, custom, date, dynamicsearch, text, textarea, year', () => {
    const computeTotalMock = jest.fn();

    const { container } = render(
      <table>
        <tbody>
          <tr>
            <StaticDataTableFooterCell<{ badge: string }>
              column={{
                dataIndex: 'badge',
                title: 'Badge',
                type: ColumnType.BADGE,
                options: [
                  { value: '1', label: '1' },
                  { value: '2', label: '2' },
                ],
              }}
              data={[{ badge: '1' }, { badge: '2' }]}
              extra={{ computeTotal: computeTotalMock }}
              rowIndex={3}
            />
          </tr>
        </tbody>
      </table>,
    );

    expect(container).toMatchSnapshot();
    expect(computeTotalMock).toBeCalledTimes(0);
  });
});
