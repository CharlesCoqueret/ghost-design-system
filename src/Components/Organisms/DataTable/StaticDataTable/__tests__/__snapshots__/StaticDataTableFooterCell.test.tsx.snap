// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StaticDataTableFooterCell component StaticDataTableFooterCell renders with amount 1`] = `
<div>
  <table>
    <tbody>
      <tr>
        <td
          class=""
        >
          <div
            class="fieldGroup"
          >
            <div
              class="default readOnly table--cell-value--amount-readonly"
            >
              <span
                class="typography"
              >
                -
              </span>
            </div>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`StaticDataTableFooterCell component StaticDataTableFooterCell renders with amount 2`] = `
<div>
  <table>
    <tbody>
      <tr>
        <td
          class=""
        >
          <div
            class="fieldGroup"
          >
            <div
              class="default readOnly table--cell-value--amount-readonly"
            >
              <span
                class="typography"
              >
                -
              </span>
            </div>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`StaticDataTableFooterCell component StaticDataTableFooterCell renders with badge, button, checkbox, code, custom, date, dynamicsearch, text, textarea, year 1`] = `
<div>
  <table>
    <tbody>
      <tr>
        <td
          class="table--footer-value center"
        >
          -
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`StaticDataTableFooterCell component StaticDataTableFooterCell renders with number 1`] = `
<div>
  <table>
    <tbody>
      <tr>
        <td
          class=""
        >
          <div
            class="fieldGroup"
          >
            <div
              class="default readOnly table--cell-value--number-readonly"
            >
              <span
                class="typography"
              >
                -
              </span>
            </div>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`StaticDataTableFooterCell component StaticDataTableFooterCell renders with number 2`] = `
<div>
  <table>
    <tbody>
      <tr>
        <td
          class=""
        >
          <div
            class="fieldGroup"
          >
            <div
              class="default readOnly table--cell-value--number-readonly"
            >
              <span
                class="typography"
              >
                -
              </span>
            </div>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`StaticDataTableFooterCell component StaticDataTableFooterCell renders with percentage 1`] = `
<div>
  <table>
    <tbody>
      <tr>
        <td
          class=""
        >
          <div
            class="fieldGroup"
          >
            <div
              class="default readOnly table--cell-value--percentage-readonly"
            >
              <span
                class="typography"
              >
                - %
              </span>
            </div>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`StaticDataTableFooterCell component StaticDataTableFooterCell renders with percentage 2`] = `
<div>
  <table>
    <tbody>
      <tr>
        <td
          class=""
        >
          <div
            class="fieldGroup"
          >
            <div
              class="default readOnly table--cell-value--percentage-readonly"
            >
              <span
                class="typography"
              >
                - %
              </span>
            </div>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
